### PR TITLE
Support tlsmode

### PR DESCRIPTION
**Pull request by sitingren**
- Add new connection option `tlsmode`.
- When connection option `ssl` is not set, change its behavior from tlsmode='disable' to tlsmode='prefer'.
- Update README [TLS/SSL](https://github.com/vertica/vertica-python?tab=readme-ov-file#tlsssl) section.

### CODE REVIEW

**Review by kevinkarch88 — APPROVED** [review]
Tests look good and thorough to me. Approving.

File: README.md [hunk 1]
```diff
@@ -106,7 +106,11 @@ with vertica_python.connect(**conn_info) as connection:
 | oauth_access_token | See [OAuth Authentication](#oauth-authentication). <br>**_Default_**: "" |
 | request_complex_types | See [SQL Data conversion to Python objects](#sql-data-conversion-to-python-objects). <br>**_Default_**: True |
 | session_label | Sets a label for the connection on the server. This value appears in the client_label column of the _v_monitor.sessions_ system table. <br>**_Default_**: an auto-generated label with format of `vertica-python-{version}-{random_uuid}` |
-| ssl | See [TLS/SSL](#tlsssl). <br>**_Default_**: False (disabled) |
+| ssl | See [TLS/SSL](#tlsssl). <br>**_Default_**: None (tlsmode="prefer") |
+| tlsmode | Controls whether the connection to the server uses TLS encryption. <br>See [TLS/SSL](#tlsssl). <br>**_Default_**: "prefer" |
+| tls_cafile | The name of a file containing SSL certificate authority (CA) certificate(s). <br>See [TLS/SSL](#tlsssl). |
```
Comment (DMickens) [inline README.md:111]: nit: I would add "trusted" in here for context. "...file containing trusted SSL certificate..."

File: README.md [hunk 2]
```diff
@@ -141,22 +145,61 @@ with vertica_python.connect(dsn=connection_str, **additional_info) as conn:
...
+There are two options to control client-server TLS: `tlsmode` and `ssl`. If both are set, `tlsmode` takes precedence.
+
+`ssl` can be a bool or a `ssl.SSLContext` object. Here is the value mapping between `ssl` (exclude `ssl.SSLContext`) and `tlsmode`:
+
+| `tlsmode` |  `ssl` | Description |
+| ------------- | ------------- | ---|
+| 'disable'     | False | only try a non-TLS connection. |
+| 'prefer'      | (not set) | (Default) first try a TLS connection; if TLS is disabled on the server, then fallback to a non-TLS connection. <br>Note: If TLS is enabled on the server and TLS connection fails, the client rejects the connection. |
+| 'require'     | True |  connects using TLS without verifying certificates. If the TLS connection attempt fails, the client rejects the connection. |
+| 'verify-ca'   || connects using TLS and confirms that the server certificate has been signed by the certificate authority. |
```
Comment (DMickens) [inline README.md:158]: similar nit, I would change the description slightly to say "...signed by a trusted certificate authority."

Comment (DMickens) [reply README.md:158]: and same change to the relevant part of verify-full

File: vertica_python/tests/integration_tests/test_tls.py [hunk 3]
```diff
@@ -84,7 +95,7 @@ def _generate_and_set_certificates(self, mutual_mode=False):
                 # This CA certificate is used to verify client certificates
                 cur.execute('ALTER TLS CONFIGURATION server CERTIFICATE vp_server_cert ADD CA CERTIFICATES vp_CA_cert')
                 # Enable TLS. Connection succeeds if Vertica verifies that the client certificate is from a trusted CA.
-                # If the client does not present a client certificate, the connection uses plaintext.
```
Comment (DMickens) [inline vertica_python/tests/integration_tests/test_tls.py:87]: good adjustment, more secure

File: vertica_python/vertica/connection.py [hunk 4]
```diff
@@ -501,11 +503,31 @@ def _socket(self):
...
-            # enable SSL
+            # enable TLS
+            tlsmode_options = self.options.get('tlsmode')
             ssl_options = self.options.get('ssl')
-            self._logger.debug('SSL option is {0}'.format('enabled' if ssl_options else 'disabled'))
-            if ssl_options:
-                raw_socket = self.enable_ssl(raw_socket, ssl_options)
+            # If TLSmode option and SSL option are set, TLSmode option takes precedence.
+            ssl_context = None
+            if tlsmode_options is not None:
```
Comment (williammatherjones) [inline vertica_python/vertica/connection.py:511]: Would this logic be clearer as a function that takes tlsmode_options and ssl_options and returns the correct tlsmode?
`tlsmode = _initTLSMode(self.options.get('tlsmode'), self.options.get('ssl'))`

File: vertica_python/vertica/tlsmode.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+import ssl
+import warnings
+from enum import Enum
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from typing import Optional
+
+
+class TLSMode(Enum):
```
Comment (williammatherjones) [inline vertica_python/vertica/tlsmode.py:26]: Nice and clean.  I like your implementation of this class.

**Review by DMickens — APPROVED** [review]
These changes look great

**Pull request merged**